# Estudando BDGD

In [1]:
import pandas as pd
import numpy as np

## Leitura: CTMT (Circuito de Média Tensão)

In [2]:
file_ctmt = './Dados/CTMT_EPB.txt'

In [3]:
# Designação da Entidade: Circuito de Média Tensão
# Designação da Modelagem: CTMT
ctmt = pd.read_csv(file_ctmt, sep=';',index_col=0)
ctmt.head(3)

COD_ID     NOM       BARR       SUB                  PAC  TEN_NOM  \
OBJECTID                                                                        
1         17067303  IBR-L1  404181344  12873436  56560487917146437MT       49   
2         17067304  IBR-L2  404181344  12873436  56560312917147011MT       49   
3         17067308  IBR-L4  273725354  12873436  56560193917147954MT       49   

          TEN_OPE  ATIP  RECONFIG  DIST  ...      PNTBT_06      PNTBT_07  \
OBJECTID                                 ...                               
1           1.029     0         0  6600  ...  25489.904739  28300.107793   
2           1.029     0         0  6600  ...  29168.047490  34798.767123   
3           1.029     0         0  6600  ...  47019.126530  54452.304372   

              PNTBT_08      PNTBT_09      PNTBT_10      PNTBT_11  \
OBJECTID                                                           
1         36189.774149  34932.604990  53273.719264  28495.406676   
2         45886.252346  43400.600957  60956.705062  33167.035131   
3         71262.602440  68976.054974  98080.409456  53571.931293   

              PNTBT_12      DESCR  PERD_A3aA4  PERD_A4A3a  
OBJECTID                                                   
1         40634.251546  SE IBIARA         0.0         0.0  
2         47549.473862  SE IBIARA         0.0         0.0  
3         78323.682392  SE IBIARA         0.0         0.0  

[3 rows x 58 columns]

In [4]:
ctmt.shape

(360, 58)

Há 58 colunas. 

Não sabemos quantas dessas colunas são de interesse, então vou excluir uma parte para ficar com uma tabela mais leve.

In [5]:
ene_cols = [f'ENE_{i:02d}' for i in range(1,13)]
ctmt = ctmt.drop(columns=ene_cols)
ctmt.head(3)

COD_ID     NOM       BARR       SUB                  PAC  TEN_NOM  \
OBJECTID                                                                        
1         17067303  IBR-L1  404181344  12873436  56560487917146437MT       49   
2         17067304  IBR-L2  404181344  12873436  56560312917147011MT       49   
3         17067308  IBR-L4  273725354  12873436  56560193917147954MT       49   

          TEN_OPE  ATIP  RECONFIG  DIST  ...      PNTBT_06      PNTBT_07  \
OBJECTID                                 ...                               
1           1.029     0         0  6600  ...  25489.904739  28300.107793   
2           1.029     0         0  6600  ...  29168.047490  34798.767123   
3           1.029     0         0  6600  ...  47019.126530  54452.304372   

              PNTBT_08      PNTBT_09      PNTBT_10      PNTBT_11  \
OBJECTID                                                           
1         36189.774149  34932.604990  53273.719264  28495.406676   
2         45886.252346  43400.600957  60956.705062  33167.035131   
3         71262.602440  68976.054974  98080.409456  53571.931293   

              PNTBT_12      DESCR  PERD_A3aA4  PERD_A4A3a  
OBJECTID                                                   
1         40634.251546  SE IBIARA         0.0         0.0  
2         47549.473862  SE IBIARA         0.0         0.0  
3         78323.682392  SE IBIARA         0.0         0.0  

[3 rows x 46 columns]

In [6]:
perd_cols = [
    'PERD_A3a',
    'PERD_A4',
    'PERD_B',
    'PERD_MED',
    'PERD_A3aA4', # O correto pelo PRODIST é 'PERD_A3a_A4'.
    'PERD_A3a_B',
    'PERD_A4A3a', # O correto pelo PRODIST é 'PERD_A4_A3a'.
    'PERD_A4_B',
    'PERD_B_A3a',
    'PERD_B_A4',
]
ctmt = ctmt.drop(columns=perd_cols)
ctmt.head(3)

COD_ID     NOM       BARR       SUB                  PAC  TEN_NOM  \
OBJECTID                                                                        
1         17067303  IBR-L1  404181344  12873436  56560487917146437MT       49   
2         17067304  IBR-L2  404181344  12873436  56560312917147011MT       49   
3         17067308  IBR-L4  273725354  12873436  56560193917147954MT       49   

          TEN_OPE  ATIP  RECONFIG  DIST  ...      PNTBT_04      PNTBT_05  \
OBJECTID                                 ...                               
1           1.029     0         0  6600  ...  16613.334788  18878.320901   
2           1.029     0         0  6600  ...  16957.582566  19803.565008   
3           1.029     0         0  6600  ...  29074.493248  33819.607247   

              PNTBT_06      PNTBT_07      PNTBT_08      PNTBT_09  \
OBJECTID                                                           
1         25489.904739  28300.107793  36189.774149  34932.604990   
2         29168.047490  34798.767123  45886.252346  43400.600957   
3         47019.126530  54452.304372  71262.602440  68976.054974   

              PNTBT_10      PNTBT_11      PNTBT_12      DESCR  
OBJECTID                                                       
1         53273.719264  28495.406676  40634.251546  SE IBIARA  
2         60956.705062  33167.035131  47549.473862  SE IBIARA  
3         98080.409456  53571.931293  78323.682392  SE IBIARA  

[3 rows x 36 columns]

In [7]:
pntmt_cols = [f'PNTMT_{i:02d}' for i in range(1,13)]
ctmt = ctmt.drop(columns=pntmt_cols)
ctmt.head(3)

COD_ID     NOM       BARR       SUB                  PAC  TEN_NOM  \
OBJECTID                                                                        
1         17067303  IBR-L1  404181344  12873436  56560487917146437MT       49   
2         17067304  IBR-L2  404181344  12873436  56560312917147011MT       49   
3         17067308  IBR-L4  273725354  12873436  56560193917147954MT       49   

          TEN_OPE  ATIP  RECONFIG  DIST  ...      PNTBT_04      PNTBT_05  \
OBJECTID                                 ...                               
1           1.029     0         0  6600  ...  16613.334788  18878.320901   
2           1.029     0         0  6600  ...  16957.582566  19803.565008   
3           1.029     0         0  6600  ...  29074.493248  33819.607247   

              PNTBT_06      PNTBT_07      PNTBT_08      PNTBT_09  \
OBJECTID                                                           
1         25489.904739  28300.107793  36189.774149  34932.604990   
2         29168.047490  34798.767123  45886.252346  43400.600957   
3         47019.126530  54452.304372  71262.602440  68976.054974   

              PNTBT_10      PNTBT_11      PNTBT_12      DESCR  
OBJECTID                                                       
1         53273.719264  28495.406676  40634.251546  SE IBIARA  
2         60956.705062  33167.035131  47549.473862  SE IBIARA  
3         98080.409456  53571.931293  78323.682392  SE IBIARA  

[3 rows x 24 columns]

In [8]:
pntbt_cols = [f'PNTBT_{i:02d}' for i in range(1,13)]
ctmt = ctmt.drop(columns=pntbt_cols)
ctmt.head(3)

COD_ID     NOM       BARR       SUB                  PAC  TEN_NOM  \
OBJECTID                                                                        
1         17067303  IBR-L1  404181344  12873436  56560487917146437MT       49   
2         17067304  IBR-L2  404181344  12873436  56560312917147011MT       49   
3         17067308  IBR-L4  273725354  12873436  56560193917147954MT       49   

          TEN_OPE  ATIP  RECONFIG  DIST  UNI_TR_S      DESCR  
OBJECTID                                                      
1           1.029     0         0  6600  12877171  SE IBIARA  
2           1.029     0         0  6600  12877171  SE IBIARA  
3           1.029     0         0  6600  12877171  SE IBIARA

**Terminou esse bloco com `ctmt` reduzido.**

## Leitura: SSDMT (Segmento de Rede Média Tensão)

In [9]:
file_ssdmt = './Dados/SSDMT_EPB.txt'

In [10]:
# Designação da Entidade: Segmento de Rede Média Tensão
# Designação da Modelagem: SSDMT
ssdmt = pd.read_csv(file_ssdmt, sep=';', index_col=0)

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3326: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [11]:
ssdmt.head(3)

COD_ID  PN_CON_1  PN_CON_2      CTMT  UNI_TR_S       SUB   CONJ  \
OBJECTID                                                                      
1         15223672  20084633  20084577  17067438  12911922  12873440  13739   
2         15223673  20083770  20084206  17067438  12911922  12873440  13739   
3         15223674  20326698  20084492  17067438  12911922  12873440  13739   

          DIST                PAC_1                PAC_2  ... ODI_FAS TI_FAS  \
OBJECTID                                                  ...                  
1         6600  58224935927088283MT  58221888927070484MT  ...       1     41   
2         6600  58191143927223327MT  58205538927212693MT  ...       1     41   
3         6600  58211624927208161MT  58216190927204185MT  ...       1     41   

         ODI_NEU TI_NEU     COMP DESCR  SITCONTFAS  SITCONTNEU ARE_LOC  \
OBJECTID                                                                 
1                     0  180.579               AT1           0      NU   
2                     0  178.969               AT1           0      NU   
3                     0   60.545               AT1           0      NU   

         Shape_Length  
OBJECTID               
1            0.001633  
2            0.001619  
3            0.000548  

[3 rows x 23 columns]

In [12]:
ssdmt.shape

(434463, 23)

Há 23 colunas. 

**Terminou esse bloco com `ssdmt` completo.**

## Obtendo a lista de SUB

In [13]:
ctmt.head(3)

COD_ID     NOM       BARR       SUB                  PAC  TEN_NOM  \
OBJECTID                                                                        
1         17067303  IBR-L1  404181344  12873436  56560487917146437MT       49   
2         17067304  IBR-L2  404181344  12873436  56560312917147011MT       49   
3         17067308  IBR-L4  273725354  12873436  56560193917147954MT       49   

          TEN_OPE  ATIP  RECONFIG  DIST  UNI_TR_S      DESCR  
OBJECTID                                                      
1           1.029     0         0  6600  12877171  SE IBIARA  
2           1.029     0         0  6600  12877171  SE IBIARA  
3           1.029     0         0  6600  12877171  SE IBIARA

In [14]:
dict_SUB = {}
for idx, row in ctmt.iterrows():
    cod_ctmt = row['COD_ID']
    sub = row['SUB']
    descr = row['DESCR']
    if sub not in dict_SUB:
        dict_SUB[sub] = {}
        dict_SUB[sub]['cod_ctmt'] = cod_ctmt
        dict_SUB[sub]['nome'] = descr
dict_SUB

{12873436: {'cod_ctmt': 17067303, 'nome': 'SE IBIARA'},
 12873433: {'cod_ctmt': 17067313, 'nome': 'SE ITAPORANGA'},
 12873430: {'cod_ctmt': 17067324, 'nome': 'SE JERICO'},
 12873438: {'cod_ctmt': 17067328, 'nome': 'SE JATOBA'},
 12912330: {'cod_ctmt': 17067341, 'nome': 'SE MALTA'},
 12873429: {'cod_ctmt': 17067350, 'nome': 'SE POMBAL'},
 12873432: {'cod_ctmt': 17067364, 'nome': 'SE PIANCO'},
 12873426: {'cod_ctmt': 17067369, 'nome': 'SE PAULISTA'},
 12873434: {'cod_ctmt': 17067376, 'nome': 'SE PRINCESA ISABEL'},
 12873439: {'cod_ctmt': 17067389, 'nome': 'SE PATOS'},
 12873428: {'cod_ctmt': 17067401, 'nome': 'SE - SÃO BENTO'},
 12912331: {'cod_ctmt': 17067410, 'nome': 'SE SAO GONÇALO'},
 12912327: {'cod_ctmt': 17067422, 'nome': 'SE SANTA LUZIA'},
 12873435: {'cod_ctmt': 17067431, 'nome': 'SE SAO JOAO DO RIO DO PEIXE'},
 12873419: {'cod_ctmt': 51696864, 'nome': 'SE ITABAIANA'},
 12873449: {'cod_ctmt': 51696866, 'nome': 'SE JACARAÚ'},
 12873483: {'cod_ctmt': 51696872, 'nome': 'SE JOÃO PES

**Terminou esse bloco com `dict_SUB` com chaves sendo os códigos de SUB.**

## Obtendo a lista de PAC's dos circuitos que se originam em cada SUB

In [15]:
ctmt.head(3)

COD_ID     NOM       BARR       SUB                  PAC  TEN_NOM  \
OBJECTID                                                                        
1         17067303  IBR-L1  404181344  12873436  56560487917146437MT       49   
2         17067304  IBR-L2  404181344  12873436  56560312917147011MT       49   
3         17067308  IBR-L4  273725354  12873436  56560193917147954MT       49   

          TEN_OPE  ATIP  RECONFIG  DIST  UNI_TR_S      DESCR  
OBJECTID                                                      
1           1.029     0         0  6600  12877171  SE IBIARA  
2           1.029     0         0  6600  12877171  SE IBIARA  
3           1.029     0         0  6600  12877171  SE IBIARA

In [16]:
for sub in dict_SUB:
    flag = (sub == ctmt['SUB'])
    ctmt_filter = ctmt[flag]
    dict_SUB[sub]['pac_circuitos'] = list(ctmt_filter['PAC'])
dict_SUB

{12873436: {'cod_ctmt': 17067303,
  'nome': 'SE IBIARA',
  'pac_circuitos': ['56560487917146437MT',
   '56560312917147011MT',
   '56560193917147954MT',
   '56560278917147543MT',
   '56561475917147463MT']},
 12873433: {'cod_ctmt': 17067313,
  'nome': 'SE ITAPORANGA',
  'pac_circuitos': ['59635125919295076MT',
   '59636090919296172MT',
   '59634679919295071MT',
   '59636086919295611MT',
   '59635129919297509MT']},
 12873430: {'cod_ctmt': 17067324,
  'nome': 'SE JERICO',
  'pac_circuitos': ['63301317927424226MT',
   '63301012927424787MT',
   '63301436927425283MT']},
 12873438: {'cod_ctmt': 17067328,
  'nome': 'SE JATOBA',
  'pac_circuitos': ['69007487922027178MT',
   '69005414922026585MT',
   '69005461922026889MT',
   '69005508922027466MT',
   '69005348922026335MT']},
 12912330: {'cod_ctmt': 17067341,
  'nome': 'SE MALTA',
  'pac_circuitos': ['66284306923656643MT',
   '66288297923655653MT',
   '66286679923658588MT',
   '66287687923655438MT']},
 12873429: {'cod_ctmt': 17067350,
  'nome': '

## Obtendo os trechos que formam os circuitos

Os trechos estão em `ssdmt`.

In [17]:
ssdmt.head(3)

COD_ID  PN_CON_1  PN_CON_2      CTMT  UNI_TR_S       SUB   CONJ  \
OBJECTID                                                                      
1         15223672  20084633  20084577  17067438  12911922  12873440  13739   
2         15223673  20083770  20084206  17067438  12911922  12873440  13739   
3         15223674  20326698  20084492  17067438  12911922  12873440  13739   

          DIST                PAC_1                PAC_2  ... ODI_FAS TI_FAS  \
OBJECTID                                                  ...                  
1         6600  58224935927088283MT  58221888927070484MT  ...       1     41   
2         6600  58191143927223327MT  58205538927212693MT  ...       1     41   
3         6600  58211624927208161MT  58216190927204185MT  ...       1     41   

         ODI_NEU TI_NEU     COMP DESCR  SITCONTFAS  SITCONTNEU ARE_LOC  \
OBJECTID                                                                 
1                     0  180.579               AT1           0      NU   
2                     0  178.969               AT1           0      NU   
3                     0   60.545               AT1           0      NU   

         Shape_Length  
OBJECTID               
1            0.001633  
2            0.001619  
3            0.000548  

[3 rows x 23 columns]

Vamos criar colunas novas para registrar subestação, circuito, de, para e contador.

In [18]:
ssdmt_com_circ = ssdmt.copy()
ssdmt_com_circ['circ_SUB'] = -1
ssdmt_com_circ['circ_CTMT'] = -1
ssdmt_com_circ['circ_PAC_de'] = -1
ssdmt_com_circ['circ_PAC_pr'] = -1
ssdmt_com_circ['circ_cont'] = -1
ssdmt_com_circ.head(3)

COD_ID  PN_CON_1  PN_CON_2      CTMT  UNI_TR_S       SUB   CONJ  \
OBJECTID                                                                      
1         15223672  20084633  20084577  17067438  12911922  12873440  13739   
2         15223673  20083770  20084206  17067438  12911922  12873440  13739   
3         15223674  20326698  20084492  17067438  12911922  12873440  13739   

          DIST                PAC_1                PAC_2  ... DESCR  \
OBJECTID                                                  ...         
1         6600  58224935927088283MT  58221888927070484MT  ...         
2         6600  58191143927223327MT  58205538927212693MT  ...         
3         6600  58211624927208161MT  58216190927204185MT  ...         

         SITCONTFAS SITCONTNEU ARE_LOC  Shape_Length circ_SUB  circ_CTMT  \
OBJECTID                                                                   
1               AT1          0      NU      0.001633       -1         -1   
2               AT1          0      NU      0.001619       -1         -1   
3               AT1          0      NU      0.000548       -1         -1   

          circ_PAC_de circ_PAC_pr circ_cont  
OBJECTID                                     
1                  -1          -1        -1  
2                  -1          -1        -1  
3                  -1          -1        -1  

[3 rows x 28 columns]

Rotina que faz a busca pelos trechos que pertencem a certo circuito.

In [19]:
def busca_em_largura(df, sub, pac_circ):
    flag = (df['circ_SUB'] == -1)
    dff = df[flag]
    
    C = [pac_circ,]
    D = set(list(dff['PAC_1']) + list(dff['PAC_2'])) - set(C)
    cont = 0

    while(len(C) > 0):

        atual = C[0]
#         print('\n\n')
#         print(f'atual = {atual}\n')
#         print('=====')
#         for item in C:
#             print(f'C.item = {item}')
#         print('=====')

        flag1 = (dff['PAC_1'] == atual)
#         print(f'sum(flag1) = {sum(flag1)}\n')

        filhos = dff[flag1]
#         display(filhos)

        for idx, row in filhos.iterrows():
            if row['PAC_2'] in D:
                print('Achou trecho!')
                df.loc[idx,'circ_SUB'] = sub
                df.loc[idx,'circ_CTMT'] = pac_circ
                df.loc[idx,'circ_PAC_de'] = row['PAC_1']
                df.loc[idx,'circ_PAC_pr'] = row['PAC_2']
                df.loc[idx,'circ_cont'] = cont
                cont += 1

                D.remove(row['PAC_2'])
                C.append(row['PAC_2'])
            else:
                print('Achou malha!')
                df.loc[idx,'circ_SUB'] = 'Malha'
                df.loc[idx,'circ_CTMT'] = 'Malha'
                df.loc[idx,'circ_PAC_de'] = 'Malha'
                df.loc[idx,'circ_PAC_pr'] = 'Malha'

        flag2 = (dff['PAC_2'] == atual)
#         print(f'sum(flag2) = {sum(flag2)}\n')
        
        filhos = dff[flag2]
#         display(filhos)
        
        for idx, row in filhos.iterrows():
            if row['PAC_1'] in D:
                print('Achou trecho!')
                df.loc[idx,'circ_SUB'] = sub
                df.loc[idx,'circ_CTMT'] = pac_circ
                df.loc[idx,'circ_PAC_de'] = row['PAC_2']
                df.loc[idx,'circ_PAC_pr'] = row['PAC_1']
                df.loc[idx,'circ_cont'] = cont
                cont += 1

                D.remove(row['PAC_1'])
                C.append(row['PAC_1'])
            else:
                print('Achou malha!')
                df.loc[idx,'circ_SUB'] = 'Malha'
                df.loc[idx,'circ_CTMT'] = 'Malha'
                df.loc[idx,'circ_PAC_de'] = 'Malha'
                df.loc[idx,'circ_PAC_pr'] = 'Malha'
    
        flag = (df['circ_SUB'] == -1)
#         print(f'sum(flag = {sum(flag)}')
        
        dff = df[flag]
        C.remove(atual)

Fazendo para apenas a 1° subestação e o 1° circuito.

In [20]:
df = ssdmt_com_circ.copy()
sub = next(iter(dict_SUB))
circ = dict_SUB[sub]['pac_circuitos'][0]
busca_em_largura(df, sub, circ)

flag = (df['circ_SUB'] != -1)
print(f'df[flag].shape = {df[flag].shape}')
print(f'df[~flag].shape = {df[~flag].shape}')
df[flag]

Achou trecho!
Achou trecho!
Achou trecho!
Achou trecho!
Achou trecho!
Achou trecho!
Achou trecho!
Achou trecho!
Achou trecho!
df[flag].shape = (9, 28)
df[~flag].shape = (434454, 28)


COD_ID   PN_CON_1   PN_CON_2      CTMT  UNI_TR_S       SUB  \
OBJECTID                                                                  
145968    344160028   20050595  344160015  17067303  12877171  12873436   
224969     15242024   20050548   20050547  17067303  12877171  12873436   
224970     15242025   20050554   20050548  17067303  12877171  12873436   
224971     15242027   12877209   20050554  17067303  12877171  12873436   
279920     15241807   20050555   20050593  17067303  12877171  12873436   
279921     15241808   20050547   20050555  17067303  12877171  12873436   
398855     15241804   20049640   20049697  17067303  12877171  12873436   
398858     15241805  344160015   20049640  17067303  12877171  12873436   
398864     15241806   20050593   20050595  17067303  12877171  12873436   

           CONJ  DIST                PAC_1                PAC_2  ... DESCR  \
OBJECTID                                                         ...         
145968    13704  6600  56561829917130593MT  56563111917130824MT  ...         
224969    13704  6600  56558719917145252MT  56559647917141454MT  ...         
224970    13704  6600  56559326917146241MT  56558719917145252MT  ...         
224971    13704  6600  56560487917146437MT  56559326917146241MT  ...         
279920    13704  6600  56560496917137638MT  56561370917133749MT  ...         
279921    13704  6600  56559647917141454MT  56560496917137638MT  ...         
398855    13704  6600  56566566917131687MT  56569593917132295MT  ...         
398858    13704  6600  56563111917130824MT  56566566917131687MT  ...         
398864    13704  6600  56561370917133749MT  56561829917130593MT  ...         

         SITCONTFAS SITCONTNEU ARE_LOC  Shape_Length  circ_SUB  \
OBJECTID                                                         
145968          AT1          0      UB      0.000118  12873436   
224969          AT1          0      UB      0.000354  12873436   
224970          AT1          0      UB      0.000105  12873436   
224971          AT1          0      UB      0.000107  12873436   
279920          AT1          0      UB      0.000361  12873436   
279921          AT1          0      UB      0.000354  12873436   
398855          AT1          0      UB      0.000280  12873436   
398858          AT1          0      UB      0.000323  12873436   
398864          AT1          0      UB      0.000288  12873436   

                    circ_CTMT          circ_PAC_de          circ_PAC_pr  \
OBJECTID                                                                  
145968    56560487917146437MT  56561829917130593MT  56563111917130824MT   
224969    56560487917146437MT  56558719917145252MT  56559647917141454MT   
224970    56560487917146437MT  56559326917146241MT  56558719917145252MT   
224971    56560487917146437MT  56560487917146437MT  56559326917146241MT   
279920    56560487917146437MT  56560496917137638MT  56561370917133749MT   
279921    56560487917146437MT  56559647917141454MT  56560496917137638MT   
398855    56560487917146437MT  56566566917131687MT  56569593917132295MT   
398858    56560487917146437MT  56563111917130824MT  56566566917131687MT   
398864    56560487917146437MT  56561370917133749MT  56561829917130593MT   

         circ_cont  
OBJECTID            
145968           6  
224969           2  
224970           1  
224971           0  
279920           4  
279921           3  
398855           8  
398858           7  
398864           5  

[9 rows x 28 columns]

Fazendo para todas as subestações e todos os circuitos.

Atenção: vou deixar comentado, porque leva uns 30 minutos.

Problema: 
>Após uns 30 minutos, temos 6813 trechos identificados.

>E 427650 trechos sem identificação. Creio que há outra fonte de trechos além de ssdmt.

Perguntar para Mikaele.

In [21]:
# df = ssdmt_com_circ.copy()
# for sub in dict_SUB:
#     for circ in dict_SUB[sub]['circuitos']:
#         busca_em_largura(df, sub, circ)
        
# flag = (df['circ_SUB'] != -1)
# print(f'df[flag].shape = {df[flag].shape}')
# print(f'df[~flag].shape = {df[~flag].shape}')

**Depuração:**

O primeiro circuito da primeira subestação ficou apenas com 9 trechos.

Vamos ver em que pac ele parou:

In [22]:
pac = '56569593917132295MT'

In [23]:
flag1 = (ssdmt_com_circ['PAC_1'] == pac)
ssdmt_com_circ[flag1]

Empty DataFrame
Columns: [COD_ID, PN_CON_1, PN_CON_2, CTMT, UNI_TR_S, SUB, CONJ, DIST, PAC_1, PAC_2, FAS_CON, TIP_CND, POS, ODI_FAS, TI_FAS, ODI_NEU, TI_NEU, COMP, DESCR, SITCONTFAS, SITCONTNEU, ARE_LOC, Shape_Length, circ_SUB, circ_CTMT, circ_PAC_de, circ_PAC_pr, circ_cont]
Index: []

[0 rows x 28 columns]

In [24]:
flag2 = (ssdmt_com_circ['PAC_2'] == pac)
ssdmt_com_circ[flag2]

COD_ID  PN_CON_1  PN_CON_2      CTMT  UNI_TR_S       SUB   CONJ  \
OBJECTID                                                                      
398855    15241804  20049640  20049697  17067303  12877171  12873436  13704   

          DIST                PAC_1                PAC_2  ... DESCR  \
OBJECTID                                                  ...         
398855    6600  56566566917131687MT  56569593917132295MT  ...         

         SITCONTFAS SITCONTNEU ARE_LOC  Shape_Length circ_SUB  circ_CTMT  \
OBJECTID                                                                   
398855          AT1          0      UB       0.00028       -1         -1   

          circ_PAC_de circ_PAC_pr circ_cont  
OBJECTID                                     
398855             -1          -1        -1  

[1 rows x 28 columns]

Observe que há apenas 1 trecho que incide nesse pac, logo o comportamento da rotina foi conforme esperado.

## Filtrando SSDMT por CTMT

No DataFrame `ssdmt` já se tem a coluna `ctmt`.

Por curiosidade, vamos buscar as ocorrências do 1° circuito da 1° subestação.

In [25]:
df = ssdmt_com_circ.copy()

sub = next(iter(dict_SUB))
cod_ctmt = dict_SUB[sub]['cod_ctmt']

flag = (ssdmt_com_circ['CTMT'] == cod_ctmt)
ssdmt_com_circ[flag]

COD_ID  PN_CON_1  PN_CON_2      CTMT  UNI_TR_S       SUB   CONJ  \
OBJECTID                                                                      
30700     15243026  20050127  20050124  17067303  12877171  12873436  13704   
30701     15243027  20050127  20050061  17067303  12877171  12873436  13704   
30702     15242986  20050140  20050166  17067303  12877171  12873436  13704   
31312     15242987  20050172  20050140  17067303  12877171  12873436  13704   
31313     15242988  20050170  20050172  17067303  12877171  12873436  13704   
...            ...       ...       ...       ...       ...       ...    ...   
398860    15241765  20204835  20204836  17067303  12877171  12873436  13704   
398861    15242090  20049361  20049360  17067303  12877171  12873436  13704   
398862    15243031  20050050  20050057  17067303  12877171  12873436  13704   
398863    15242055  20049358  20049357  17067303  12877171  12873436  13704   
398864    15241806  20050593  20050595  17067303  12877171  12873436  13704   

          DIST                PAC_1                PAC_2  ... DESCR  \
OBJECTID                                                  ...         
30700     6600  56538253917035638MT  56536395917033146MT  ...         
30701     6600  56538253917035638MT  56539430917034611MT  ...         
30702     6600  56558187917014669MT  56561520917012638MT  ...         
31312     6600  56553578917016909MT  56558187917014669MT  ...         
31313     6600  56549619917018940MT  56553578917016909MT  ...         
...        ...                  ...                  ...  ...   ...   
398860    6600  56628037917055752MT  56640739917061961MT  ...         
398861    6600  56508243917042831MT  56504367917045094MT  ...         
398862    6600  56540296917024149MT  56543895917022056MT  ...         
398863    6600  56521864917034533MT  56522561917035498MT  ...         
398864    6600  56561370917133749MT  56561829917130593MT  ...         

         SITCONTFAS SITCONTNEU ARE_LOC  Shape_Length circ_SUB  circ_CTMT  \
OBJECTID                                                                   
30700           AT1          0      UB      0.000281       -1         -1   
30701           AT1          0      UB      0.000155       -1         -1   
30702           AT1          0      UB      0.000354       -1         -1   
31312           AT1          0      UB      0.000464       -1         -1   
31313           AT1          0      UB      0.000403       -1         -1   
...             ...        ...     ...           ...      ...        ...   
398860          AT1          0      NU      0.001295       -1         -1   
398861          AT1          0      UB      0.000407       -1         -1   
398862          AT1          0      UB      0.000377       -1         -1   
398863          AT1          0      UB      0.000122       -1         -1   
398864          AT1          0      UB      0.000288       -1         -1   

          circ_PAC_de circ_PAC_pr circ_cont  
OBJECTID                                     
30700              -1          -1        -1  
30701              -1          -1        -1  
30702              -1          -1        -1  
31312              -1          -1        -1  
31313              -1          -1        -1  
...               ...         ...       ...  
398860             -1          -1        -1  
398861             -1          -1        -1  
398862             -1          -1        -1  
398863             -1          -1        -1  
398864             -1          -1        -1  

[146 rows x 28 columns]

A rotina automática pegou 9 trechos.

Esse filtro pegou 146 trechos.

Por quê?